ssl/quic/quic_channel.c

```diff
@@ -480,6 +484,37 @@ ossl_quic_channel_get_terminate_cause(const QUIC_CHANNEL *ch)
     return ossl_quic_channel_is_term_any(ch) ? &ch->terminate_cause : NULL;
 }
 
+int ossl_quic_channel_can_send_closing_bytes(QUIC_CHANNEL *ch, size_t n)
+{
+    int res;
+
+    /* We can always send the first connection close frame */
+    if (ch->closing_bytes_recv == 0)
+        return 1;
+
+    /*
+     * RFC 9000 s. 10.2.1 Closing Connection State:
+     *      To avoid being used for an amplification attack, such
+     *      endpoints MUST limit the cumulative size of packets it sends
+     *      to three times the cumulative size of the packets that are
+     *      received and attributed to the connection.
+     * and:
+     *      An endpoint in the closing state MUST either discard packets
+     *      received from an unvalidated address or limit the cumulative
+     *      size of packets it sends to an unvalidated address to three
+     *      times the size of packets it receives from that address.
+     */
+    res = ch->closing_bytes_xmit + n <= ch->closing_bytes_recv * 3;
+
+    /*
+     * Attribute the bytes to the connection, if we are allowed to send them
+     * and this isn't the first closing frame.
+     */
+    if (res && ch->closing_bytes_recv != 0)
+        ch->closing_bytes_xmit += n;
```

Comment:
It is a bit confusing to have a function whose name implies it's a non-mutating predicate but which does a mutation like this. Maybe `ossl_quic_channel_try_commit_conn_close` or something better.

There's a question as to whether this logic should live in the TXP but I can live with it being like this for now (potential issues with coupling the TXP to the RX side also).

Reply:
I've moved the logic to the TXP.  It probably sits more nicely there, although it's not quite so clean IMO.